### PR TITLE
JAX: Add AlexNet

### PR DESCRIPTION
**Pull request by AnirudhDagar**
*Description of changes:*

This PR updates the D2L API, enabling `jax.jit` for the `train`/`val` step. This speeds up the training process a lot, and in the future, we can potentially further improve this by targetting jit to a larger block, giving XLA more opportunity to optimize, if the D2L API becomes less of a constraint.

Also, it would be nice todo to use jax profiler to understand and analyse memory and time across the ModernCNN networks.

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

File: chapter_linear-classification/classification.md [hunk 1]
```diff
@@ -130,6 +132,7 @@ def accuracy(self, Y_hat, Y, averaged=True):
 ```{.python .input  n=9}
 %%tab jax
 @d2l.add_to_class(Classifier)  #@save
+@partial(jax.jit, static_argnums=(0,4))
```
Comment (astonzhang) [inline chapter_linear-classification/classification.md:135]: add a space right after comma

File: chapter_linear-classification/softmax-regression-concise.md [hunk 2]
```diff
@@ -190,8 +191,9 @@ def loss(self, Y_hat, Y, averaged=True):
 ```{.python .input}
 %%tab jax
 @d2l.add_to_class(d2l.Classifier)  #@save
-def loss(self, params, X, Y, averaged=True, rngs=None):
-    Y_hat = self.apply(params, X, rngs=rngs)
+@partial(jax.jit, static_argnums=(0,4))
```
Comment (astonzhang) [inline chapter_linear-classification/softmax-regression-concise.md:194]: add a space right after comma

File: chapter_multilayer-perceptrons/dropout.md [hunk 3]
```diff
@@ -360,6 +358,14 @@ according to the specified dropout probability.
 When not in training mode,
 the `Dropout` layer simply passes the data through during testing.
 
+:begin_tab:`jax`
+Note that we need to redefine the loss function since a network
+with a dropout layer needs a PRNGKey when using `Module.apply()`,
+this RNG seed should be explicitly named `dropout`. This key is
```
Comment (astonzhang) [inline chapter_multilayer-perceptrons/dropout.md:364]: this RNG XXX ->
and this RNG XXX

(if you prefer a long sentence here)

File: chapter_multilayer-perceptrons/dropout.md [hunk 4]
```diff
@@ -423,10 +429,17 @@ class DropoutMLP(d2l.Classifier):
         x = nn.relu(nn.Dense(self.num_hiddens_2)(x))
         x = nn.Dropout(self.dropout_2, deterministic=not self.training)(x)
         return nn.Dense(self.num_outputs)(x)
+```
```
Comment (astonzhang) [inline chapter_multilayer-perceptrons/dropout.md:432]: Note that if there's no text cell in between two code cells, both code cells will be merged during compilation.

File: chapter_multilayer-perceptrons/dropout.md [hunk 5]
```diff
@@ -423,10 +429,17 @@ class DropoutMLP(d2l.Classifier):
...
 
-@d2l.add_to_class(DropoutMLP)
-def loss(self, params, X, Y, averaged=True, rngs={'dropout': d2l.get_key()}):
-    return super(DropoutMLP, self).loss(params, X, Y, averaged, rngs)
+```{.python .input}
+%%tab jax
+@d2l.add_to_class(d2l.Classifier)  #@save
+@partial(jax.jit, static_argnums=(0,4))
```
Comment (astonzhang) [inline chapter_multilayer-perceptrons/dropout.md:437]: space

File: chapter_multilayer-perceptrons/dropout.md [hunk 6]
```diff
@@ -423,10 +429,17 @@ class DropoutMLP(d2l.Classifier):
...
+def loss(self, params, X, Y, averaged=True):
```
Comment (astonzhang) [inline chapter_multilayer-perceptrons/dropout.md:438]: You seem to overwrite Classifier.loss. 
Do we still need to keep the `averaged` argument?

Comment (AnirudhDagar) [reply chapter_multilayer-perceptrons/dropout.md:438]: Yes, as mentioned in the note the only difference is that we pass a specific rng for dropout which helps flax to generate the dropout mask. This is required in training mode for flax during `model.apply()`

File: chapter_multilayer-perceptrons/dropout.md (hunk 4 above)
Comment (AnirudhDagar) [reply chapter_multilayer-perceptrons/dropout.md:432]: Nice catch, I'll just move the jax note below.

**Comment by AnirudhDagar** [comment]
@astonzhang should be ready to merge now!

**Comment by astonzhang** [comment]
Good job on jit! Thanks
